velox/connectors/hive/storage_adapters/s3fs/S3Util.cpp

```diff
@@ -43,4 +44,107 @@ std::string getErrorStringFromS3Error(
   }
 }
 
+/// The noProxyList is a comma separated list of subdomains or domains.
+/// For a given hostname check if it has a matching domain or subdomain in
+/// the noProxyList.
+bool isHostExcludedFromProxy(
+    const std::string& hostname,
+    const std::string& noProxyList) {
+  std::vector<std::string> noProxyListElements{};
+
+  if (noProxyList.empty()) {
+    return false;
+  }
+
+  folly::split(',', noProxyList, noProxyListElements);
+  // An exact match or subdomain match is needed.
+  for (auto elem : noProxyListElements) {
+    if (elem.length() < hostname.length() && elem[0] == '.' &&
+        !hostname.compare(
+            hostname.length() - elem.length(), elem.length(), elem)) {
+      return true;
+    } else if (
+        elem.length() < hostname.length() && elem[0] == '*' && elem[1] == '.' &&
+        !hostname.compare(
+            hostname.length() - elem.length() + 1,
+            elem.length() - 1,
+            elem.substr(1))) {
+      return true;
+    } else if (elem.length() == hostname.length() && !hostname.compare(elem)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Reading the various proxy related environment variables.
+/// There is a lacking standard. The environment variables can be
+/// defined lower case or upper case. The lower case values are checked
+/// first and, if set, returned, therefore taking precendence.
+namespace {
+std::string readProxyEnvVar(std::string envVar) {
+  std::string result;
+  folly::toLowerAscii(envVar);
+  auto httpProxy = getenv(envVar.c_str());
+  if (httpProxy) {
+    result = std::string(httpProxy);
+    return result;
+  }
+
+  std::transform(envVar.begin(), envVar.end(), envVar.begin(), ::toupper);
+  httpProxy = getenv(envVar.c_str());
+  if (httpProxy) {
+    result = std::string(httpProxy);
+    return result;
+  }
+  return result;
+};
+} // namespace
+
+std::string getHttpProxyEnvVar() {
+  return readProxyEnvVar("http_proxy");
+}
+
+std::string getHttpsProxyEnvVar() {
+  return readProxyEnvVar("https_proxy");
+};
+
+std::string getNoProxyEnvVar() {
+  return readProxyEnvVar("no_proxy");
+};
+
+S3ProxyConfiguration S3ProxyConfigurationBuilder::build() {
+  S3ProxyConfiguration proxyConfig;
```

Comment:
Should we use `std::optional<S3ProxyConfiguration>` to indicate that the builder can return an empty config?

Reply:
Why not return `std::optional<folly::Uri>`? We can avoid  defining `S3ProxyConfiguration`.

Reply:
Yup, should have realized this as well. In the end we just parse the URI.